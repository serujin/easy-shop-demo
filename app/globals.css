@import url("https://p.typekit.net/p.css?s=1&k=gws6bus&ht=tk&f=139&a=10612664&app=typekit&e=css");
@font-face {
  font-family: "proxima-nova";
  src: url("https://use.typekit.net/af/949f99/00000000000000003b9b3068/27/l?primer=7cdcb44be4a7db8877ffa5c0007b8dd865b3bbc383831fe2ea177f62257a9191&fvd=n7&v=3")
      format("woff2"),
    url("https://use.typekit.net/af/949f99/00000000000000003b9b3068/27/d?primer=7cdcb44be4a7db8877ffa5c0007b8dd865b3bbc383831fe2ea177f62257a9191&fvd=n7&v=3")
      format("woff"),
    url("https://use.typekit.net/af/949f99/00000000000000003b9b3068/27/a?primer=7cdcb44be4a7db8877ffa5c0007b8dd865b3bbc383831fe2ea177f62257a9191&fvd=n7&v=3")
      format("opentype");
  font-display: auto;
  font-style: normal;
  font-weight: 700;
  font-stretch: normal;
}

* {
  box-sizing: border-box;
  padding: 0;
  margin: 0;
  font-family: "proxima-nova";
}

html {
  color: #404040;
  font-size: 14px;
  background-color: #efefef;
}

body {
  display: flex;
  flex-direction: column;
}

header {
  display: flex;
  background-color: #4e2817;
  justify-content: center;
}

a {
  text-decoration: none;
  color: black;
}
