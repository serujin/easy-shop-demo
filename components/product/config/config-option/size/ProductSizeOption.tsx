import React, { ChangeEvent } from "react";
import styles from "./ProductSizeOption.module.css";

export type ProductSizeOptionType = {
  value: string;
};

type ProductSizeOptionProps = ProductSizeOptionType & {
  handleSizeChange: (event: ChangeEvent<HTMLInputElement>) => void;
  optionNumber: number;
  selectedSize: string;
};

const createId = (optionNumber: number, value: string) => {
  const dimension = value
    .replace(/[^\w\s]/gi, "")
    .toLowerCase()
    .replace(" ", "-");
  return `size-${dimension}-${optionNumber}`;
};

export default function ProductSizeOption(props: ProductSizeOptionProps) {
  const { handleSizeChange, optionNumber, selectedSize, value } = props;
  const id = createId(optionNumber, value);
  return (
    <li className={styles.productSizeOptionContainer}>
      <input
        className={styles.productSizeOptionInput}
        type="radio"
        id={id}
        name="dimensions"
        value={value}
        checked={selectedSize === value}
        onChange={handleSizeChange}
      />
      <label className={styles.productSizeOptionLabel} htmlFor={id}>
        {value}
      </label>
    </li>
  );
}
