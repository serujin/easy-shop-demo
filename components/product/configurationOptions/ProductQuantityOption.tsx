import React, { ChangeEvent } from "react";
import styles from "./ProductQuantityOption.module.css";

export type ProductQuantityOptionType = {
  discount?: number;
  price?: Amount;
  value: string;
  isLast?: boolean;
};

type ProductQuantityOptionProps = ProductQuantityOptionType & {
  handleQuantityChange: (event: ChangeEvent<HTMLInputElement>) => void;
  optionNumber: number;
  selectedQuantity: string;
};

const createId = (optionNumber: number, value: string) =>
  `quantity-${value}-${optionNumber}`;

export default function ProductQuantityOption(
  props: ProductQuantityOptionProps
) {
  const {
    discount,
    handleQuantityChange,
    optionNumber,
    price,
    selectedQuantity,
    value,
    isLast = false,
  } = props;

  const id = createId(optionNumber, value);
  const discountLabel = discount ? `Save ${discount}%` : "";
  const priceLabelText = price ? `${price.quantity}${price.currency}` : "";

  return (
    <tr className={styles.quantityRow}>
      <td className={styles.inputCell}>
        <input
          type="radio"
          id={id}
          name="quantities"
          value={value}
          checked={selectedQuantity === value}
          onChange={handleQuantityChange}
        />
      </td>
      {isLast && (
        <td colSpan={3}>
          <label htmlFor={id}>{value}</label>
        </td>
      )}
      {!isLast && (
        <td>
          <label htmlFor={id}>{value}</label>
        </td>
      )}
      {price && (
        <td className={styles.rightAlign}>
          <label htmlFor={id}>{priceLabelText}</label>
        </td>
      )}
      {discount && (
        <td className={styles.rightAlign}>
          <p>{discountLabel}</p>
        </td>
      )}
    </tr>
  );
}
