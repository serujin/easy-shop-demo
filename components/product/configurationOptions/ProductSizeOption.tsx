import React, { ChangeEvent } from "react";

export type ProductSizeOptionType = {
  value: string;
};

type ProductSizeOptionProps = ProductSizeOptionType & {
  handleSizeChange: (event: ChangeEvent<HTMLInputElement>) => void;
  optionNumber: number;
  selectedSize: string;
};

const createId = (optionNumber: number, value: string) => {
  const dimension = value.replace(/[^\dx]/gi, "").toLowerCase();
  return `size-${dimension}-${optionNumber}`;
};

export default function ProductSizeOption(props: ProductSizeOptionProps) {
  const { handleSizeChange, optionNumber, selectedSize, value } = props;
  const id = createId(optionNumber, value);
  return (
    <li>
      <input
        type="radio"
        id={id}
        name="dimensions"
        value={value}
        checked={selectedSize === value}
        onChange={handleSizeChange}
      />
      <label htmlFor={id}>{value}</label>
    </li>
  );
}
