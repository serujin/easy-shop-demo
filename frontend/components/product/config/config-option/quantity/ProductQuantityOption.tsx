import React, { ChangeEvent } from 'react'
import styles from './ProductQuantityOption.module.css'

export type ProductQuantityOptionType = {
  discount?: number
  price?: Amount
  value: string
  isLast?: boolean
}

type ProductQuantityOptionProps = ProductQuantityOptionType & {
  handleQuantityChange: (event: ChangeEvent<HTMLInputElement>) => void
  optionNumber: number
  selectedQuantity: string
}

const getId = (optionNumber: number, value: string) => {
  const valueId = value.toLowerCase().replace(' ', '-')
  return `quantity-${valueId}-${optionNumber}`
}

const getDiscountLabel = (discount?: number) =>
  discount ? `Save ${discount}%` : ''

const getPriceLabel = (price?: Amount) =>
  price ? `${price.currency}${price.quantity}` : ''

const getValueTableCell = (isLast: boolean, id: string, value: string) =>
  isLast ? (
    <td colSpan={3}>
      <label htmlFor={id}>{value}</label>
    </td>
  ) : (
    <td>
      <label htmlFor={id}>{value}</label>
    </td>
  )

export default function ProductQuantityOption(
  props: ProductQuantityOptionProps,
) {
  const {
    discount,
    handleQuantityChange,
    optionNumber,
    price,
    selectedQuantity,
    value,
    isLast = false,
  } = props

  const id = getId(optionNumber, value)
  const discountLabel = getDiscountLabel(discount)
  const priceLabelText = getPriceLabel(price)
  const discountTableCellClassName = `${styles.rightAlign} ${styles.discount}`

  return (
    <tr className={styles.quantityRow}>
      <td className={styles.inputCell}>
        <input
          type="radio"
          id={id}
          name="quantities"
          value={value}
          checked={selectedQuantity === value}
          onChange={handleQuantityChange}
        />
      </td>
      {getValueTableCell(isLast, id, value)}
      {price && (
        <td className={styles.rightAlign}>
          <label htmlFor={id}>{priceLabelText}</label>
        </td>
      )}
      {discount && (
        <td className={discountTableCellClassName}>
          <label htmlFor={id}>{discountLabel}</label>
        </td>
      )}
    </tr>
  )
}
