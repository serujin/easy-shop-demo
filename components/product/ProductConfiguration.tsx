import React, { ChangeEvent, FormEvent, useState } from "react";
import ProductSizeOption, {
  ProductSizeOptionType,
} from "./configurationOptions/ProductSizeOption";
import ProductQuantityOption, {
  ProductQuantityOptionType,
} from "./configurationOptions/ProductQuantityOption";

type ProductConfigurationProps = {
  sizeOptions: Array<ProductSizeOptionType>;
  quantityOptions: Array<ProductQuantityOptionType>;
};

export default function ProductConfiguration(props: ProductConfigurationProps) {
  const [selectedSize, setSelectedSize] = useState<string>("");
  const [selectedQuantity, setSelectedQuantity] = useState<string>("");

  const { sizeOptions, quantityOptions } = props;

  const handleSizeChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSelectedSize(event.target.value);
  };

  const handleQuantityChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSelectedQuantity(event.target.value);
  };

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault();
    // Handle form submission here
  };

  return (
    <form noValidate onSubmit={handleSubmit}>
      <div>
        <h4>Select a size</h4>
        <ul>
          {sizeOptions &&
            sizeOptions.map(({ value }, index) => (
              <ProductSizeOption
                handleSizeChange={handleSizeChange}
                key={index}
                optionNumber={index}
                selectedSize={selectedSize}
                value={value}
              />
            ))}
        </ul>
      </div>
      <div>
        <h4>Select a quantity</h4>
        <ul>
          {quantityOptions &&
            quantityOptions.map(({ discount, price, value }, index) => (
              <ProductQuantityOption
                discount={discount}
                handleQuantityChange={handleQuantityChange}
                key={index}
                optionNumber={index}
                price={price}
                selectedQuantity={selectedQuantity}
                value={value}
              />
            ))}
        </ul>
      </div>
      <button type="submit">Continue</button>
    </form>
  );
}
