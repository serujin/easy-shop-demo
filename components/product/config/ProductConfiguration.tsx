import React, { ChangeEvent, FormEvent, useState } from 'react'
import styles from './ProductConfiguration.module.css'
import ProductSizeOption, {
  ProductSizeOptionType,
} from './config-option/size/ProductSizeOption'
import ProductQuantityOption, {
  ProductQuantityOptionType,
} from './config-option/quantity/ProductQuantityOption'

type ProductConfigurationProps = {
  sizeOptions: Array<ProductSizeOptionType>
  quantityOptions: Array<ProductQuantityOptionType>
}

export default function ProductConfiguration(props: ProductConfigurationProps) {
  const [selectedSize, setSelectedSize] = useState<string>('')
  const [selectedQuantity, setSelectedQuantity] = useState<string>('')

  const { sizeOptions, quantityOptions } = props
  const handleSizeChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSelectedSize(event.target.value)
  }

  const handleQuantityChange = (event: ChangeEvent<HTMLInputElement>) => {
    setSelectedQuantity(event.target.value)
  }

  const handleSubmit = (event: FormEvent<HTMLFormElement>) => {
    event.preventDefault()
    console.log(`Ordering ${selectedQuantity} stickers of size ${selectedSize}`)
  }

  const renderSelectedSize =
    selectedSize === '' ? sizeOptions?.at(0)?.value ?? '' : selectedSize
  const renderSelectedQuantity =
    selectedQuantity === ''
      ? quantityOptions?.at(0)?.value ?? ''
      : selectedQuantity

  return (
    <form noValidate onSubmit={handleSubmit}>
      <div className={styles.sizeOptionsContainer}>
        <h4>Select a size</h4>
        <ul>
          {sizeOptions &&
            sizeOptions.map(({ value }, index) => (
              <ProductSizeOption
                handleSizeChange={handleSizeChange}
                key={index}
                optionNumber={index}
                selectedSize={renderSelectedSize}
                value={value}
              />
            ))}
          <ProductSizeOption
            handleSizeChange={handleSizeChange}
            optionNumber={sizeOptions ? sizeOptions.length : 1}
            selectedSize={renderSelectedSize}
            value={'Custom size'}
          />
        </ul>
      </div>
      <div className={styles.quantityOptionsContainer}>
        <h4>Select a quantity</h4>
        <table>
          <tbody>
            {quantityOptions &&
              quantityOptions.map(({ discount, price, value }, index) => (
                <ProductQuantityOption
                  discount={discount}
                  handleQuantityChange={handleQuantityChange}
                  key={index}
                  optionNumber={index}
                  price={price}
                  selectedQuantity={renderSelectedQuantity}
                  value={value}
                />
              ))}
            <ProductQuantityOption
              handleQuantityChange={handleQuantityChange}
              optionNumber={quantityOptions ? quantityOptions.length : 1}
              selectedQuantity={renderSelectedQuantity}
              value={'Custom quantity'}
            />
          </tbody>
        </table>
      </div>
      <div className={styles.submitButtonContainer}>
        <button type="submit">Continue</button>
        <p>Next: upload artwork â†’</p>
      </div>
    </form>
  )
}
