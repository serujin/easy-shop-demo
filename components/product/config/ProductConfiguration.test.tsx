import React from "react";
import { render } from "@testing-library/react";
import ProductConfiguration from "./ProductConfiguration";
import { describe, expect, it } from "@jest/globals";
import { ProductQuantityOptionType } from "./config-option/quantity/ProductQuantityOption";
import { ProductSizeOptionType } from "./config-option/size/ProductSizeOption";

const mockAmount = { quantity: 0, currency: "$" };
const mockQuantityOptions: ProductQuantityOptionType[] = [
  { value: "value-0", price: mockAmount },
  { value: "value-1", discount: 20, price: mockAmount },
  { value: "value-2", discount: 30, price: mockAmount },
  { value: "value-3", discount: 40, price: mockAmount },
  { value: "value-4", discount: 50, price: mockAmount },
  { value: "value-5", isLast: true },
];
const mockSizeOptions: ProductSizeOptionType[] = [
  { value: "value-0" },
  { value: "value-1" },
  { value: "value-2" },
  { value: "value-3" },
  { value: "value-4" },
  { value: "value-5" },
];

describe("ProductConfiguration component", () => {
  it("should render correctly", () => {
    const { container } = render(
      <ProductConfiguration
        quantityOptions={mockQuantityOptions}
        sizeOptions={mockSizeOptions}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
