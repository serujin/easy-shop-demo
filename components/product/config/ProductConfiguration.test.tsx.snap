// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductConfiguration component should render correctly 1`] = `
<div>
  <form
    novalidate=""
  >
    <div
      class="sizeOptionsContainer"
    >
      <h4>
        Select a size
      </h4>
      <ul>
        <li
          class="productSizeOptionContainer"
        >
          <input
            checked=""
            class="productSizeOptionInput"
            id="size-value0-0"
            name="dimensions"
            type="radio"
            value="value-0"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value0-0"
          >
            value-0
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-value1-1"
            name="dimensions"
            type="radio"
            value="value-1"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value1-1"
          >
            value-1
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-value2-2"
            name="dimensions"
            type="radio"
            value="value-2"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value2-2"
          >
            value-2
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-value3-3"
            name="dimensions"
            type="radio"
            value="value-3"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value3-3"
          >
            value-3
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-value4-4"
            name="dimensions"
            type="radio"
            value="value-4"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value4-4"
          >
            value-4
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-value5-5"
            name="dimensions"
            type="radio"
            value="value-5"
          />
          <label
            class="productSizeOptionLabel"
            for="size-value5-5"
          >
            value-5
          </label>
        </li>
        <li
          class="productSizeOptionContainer"
        >
          <input
            class="productSizeOptionInput"
            id="size-custom-size-6"
            name="dimensions"
            type="radio"
            value="Custom size"
          />
          <label
            class="productSizeOptionLabel"
            for="size-custom-size-6"
          >
            Custom size
          </label>
        </li>
      </ul>
    </div>
    <div
      class="quantityOptionsContainer"
    >
      <h4>
        Select a quantity
      </h4>
      <table>
        <tbody>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                checked=""
                id="quantity-value-0-0"
                name="quantities"
                type="radio"
                value="value-0"
              />
            </td>
            <td>
              <label
                for="quantity-value-0-0"
              >
                value-0
              </label>
            </td>
            <td
              class="rightAlign"
            >
              <label
                for="quantity-value-0-0"
              >
                $0
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-value-1-1"
                name="quantities"
                type="radio"
                value="value-1"
              />
            </td>
            <td>
              <label
                for="quantity-value-1-1"
              >
                value-1
              </label>
            </td>
            <td
              class="rightAlign"
            >
              <label
                for="quantity-value-1-1"
              >
                $0
              </label>
            </td>
            <td
              class="rightAlign discount"
            >
              <label
                for="quantity-value-1-1"
              >
                Save 20%
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-value-2-2"
                name="quantities"
                type="radio"
                value="value-2"
              />
            </td>
            <td>
              <label
                for="quantity-value-2-2"
              >
                value-2
              </label>
            </td>
            <td
              class="rightAlign"
            >
              <label
                for="quantity-value-2-2"
              >
                $0
              </label>
            </td>
            <td
              class="rightAlign discount"
            >
              <label
                for="quantity-value-2-2"
              >
                Save 30%
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-value-3-3"
                name="quantities"
                type="radio"
                value="value-3"
              />
            </td>
            <td>
              <label
                for="quantity-value-3-3"
              >
                value-3
              </label>
            </td>
            <td
              class="rightAlign"
            >
              <label
                for="quantity-value-3-3"
              >
                $0
              </label>
            </td>
            <td
              class="rightAlign discount"
            >
              <label
                for="quantity-value-3-3"
              >
                Save 40%
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-value-4-4"
                name="quantities"
                type="radio"
                value="value-4"
              />
            </td>
            <td>
              <label
                for="quantity-value-4-4"
              >
                value-4
              </label>
            </td>
            <td
              class="rightAlign"
            >
              <label
                for="quantity-value-4-4"
              >
                $0
              </label>
            </td>
            <td
              class="rightAlign discount"
            >
              <label
                for="quantity-value-4-4"
              >
                Save 50%
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-value-5-5"
                name="quantities"
                type="radio"
                value="value-5"
              />
            </td>
            <td>
              <label
                for="quantity-value-5-5"
              >
                value-5
              </label>
            </td>
          </tr>
          <tr
            class="quantityRow"
          >
            <td
              class="inputCell"
            >
              <input
                id="quantity-custom-quantity-6"
                name="quantities"
                type="radio"
                value="Custom quantity"
              />
            </td>
            <td>
              <label
                for="quantity-custom-quantity-6"
              >
                Custom quantity
              </label>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div
      class="submitButtonContainer"
    >
      <button
        type="submit"
      >
        Continue
      </button>
      <p>
        Next: upload artwork â†’
      </p>
    </div>
  </form>
</div>
`;
